propolis/src/inventory.rs

```diff
@@ -75,27 +75,22 @@ impl Inventory {
     pub fn register<T: Entity>(
         &self,
         ent: &Arc<T>,
-        instance_name: Option<String>,
-        parent_id: Option<EntityID>,
     ) -> Result<EntityID, RegistrationError> {
         let mut inv = self.inner.lock().unwrap();
-        let to_register = ent.child_register();
-        let res = inv.register(ent, instance_name, parent_id)?;
+        inv.register(ent, None)
+    }
 
-        if let Some(children) = to_register {
-            for child in children {
-                // Since the parent successfully registered, the children should
-                // have no issues.
-                inv.register_inner(
-                    child.ent,
-                    child.ent_any,
-                    child.instance_name,
-                    Some(res),
-                )
-                .unwrap();
-            }
-        }
-        Ok(res)
+    /// Registers a new entity, bearing an instance name, with the inventory.
+    ///
+    /// Returns an error if the object has already been registered.
+    /// Returns an error if the parent is not registered.
+    pub fn register_instance<T: Entity>(
```

Comment:
It looks like you basically unnested a bunch of devices as part of using this method but not all of them are implicitly registered in their (former) parent's `child_register` method.

Reply:
`register_child` was left for cases such as block devices where the entity does not hold a reference to the child when it itself is being added to the registry.

I'll remove the note about the missing parent, since that error shouldn't apply here for now.